src/ripple/app/consensus/RCLValidations.cpp

```diff
@@ -240,24 +240,45 @@ RCLValidations::onStale(RCLValidation&& v)
         return;
 
     writing_ = true;
-    app_.getJobQueue().addJob(
-        jtWRITE, "Validations::doWrite", [this](Job&) { doWrite(); });
+    app_.getJobQueue().addJob(jtWRITE, "Validations::doWrite", [this](Job&) {
+
+        auto event =
+            app_.getJobQueue().getLoadEventAP(jtDISK, "ValidationWrite");
+        ScopedLockType sl(lock_);
+        doWrite(sl);
+    });
 }
 
 void
 RCLValidations::flush()
 {
     JLOG(j_.info()) << "Flushing validations";
-    ScopedLockType sl(lock_);
+    bool anyNew = false;
+    {
+        ScopedLockType sl(lock_);
 
-    Validations::flush();
+        Validations::flush([&](RCLValidation && v)
```

Comment:
You may be looking at the code more closely than I am, but I'm not seeing any data races if the lock is inside the lambda.  `anyNew` is a local, so it is not subject to thread races.

If we put the lock inside the lambda, we also need to add a lock just in front of line 268 to protect the rest of the function.  If that's what you're worried about then I agree.  We need a lock there as well.

In terms of what is "outside code" my personal approach is to have these buckets:
- `std` library code.
- non-`std` library, but well vetted, library code (like boost and beast).
- code in this same class.
- everything else (including other functions/classes in this code base).

Most calls into `std` library code I don't worry about.  The characteristics of that code are stable and well understood.  I get more nervous as I go further down the list because the code is less well characterized and more volatile.  Note that, in my opinion, calling into our own code outside this class is the scariest.  It is the most subject to change and we have the smallest understanding of the consequences of these changes.

Let me provide some more background for my concern.  When I hold a lock and then call an outside (non-`std` library) function I am, in effect, saying two things:
- It is okay if I hold this lock indefinitely, since I don't know how long this lock will be held.
- I strongly believe that there will be no lock order problems introduced by the outside code trying to acquire another lock or to re-acquire this lock.

I believe that you've checked all of the code that you're modifying at this moment.  I'm sure it's all good right now.

My primary concern is that, as the outside code is maintained, those maintainers have no idea that this lock is being held.  So problems like the ones I just described are easy to introduce during maintenance.

As I said, the existing code base is pretty bad (in my opinion) about holding locks over external calls.  And sometimes holding a lock over an external call is just flat-out unavoidable.  So you'll not be doing anything that isn't seen elsewhere in the code base.  I suspect in this case it *is* avoidable (at the cost of some extra lock thrash).  I'm just pointing it out for your consideration.

You know best how you should manage the lock, and I leave that to your discretion.  I'm sure you'll make the best choice for the code base.  Thanks for listening.